File: src/main/java/org/distril/beengine/world/util/Location.java
package org.distril.beengine.world.util;

import com.nukkitx.math.vector.Vector3f;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.distril.beengine.world.World;
import org.distril.beengine.world.chunk.Chunk;

import java.util.Objects;

@Getter
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class Location {

	private final Vector3f position;
	private final World world;

	public static Location from(Vector3f position, World world) {
		return new Location(position, world);
	}

	public static Location from(World world) {
		return Location.from(Vector3f.ZERO, world);
	}

	public float getX() {
		return this.position.getX();
	}

	public float getY() {
		return this.position.getY();
	}

	public float getZ() {
		return this.position.getZ();
	}

	public int getFloorX() {
		return this.position.getFloorX();
	}

	public int getFloorY() {
		return this.position.getFloorY();
	}

	public int getFloorZ() {
		return this.position.getFloorZ();
	}

	public Chunk getChunk() {
		return this.world.getChunk(this.getChunkX(), this.getChunkZ());
	}

	public int getChunkX() {
		return this.getFloorX() >> 4;
	}

	public int getChunkZ() {
		return this.getFloorZ() >> 4;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}

		if (obj == null || this.getClass() != obj.getClass()) {
			return false;
		}

		Location that = (Location) obj;
		return this.position.equals(that.getPosition()) && this.world.equals(that.getWorld());
	}

	@Override
	public int hashCode() {
		return Objects.hash(this.position, this.world);
	}
}

File: src/main/java/org/distril/beengine/world/util/Location.kt
package org.distril.beengine.world.util

import com.nukkitx.math.vector.Vector3f
import org.distril.beengine.world.World
import org.distril.beengine.world.chunk.Chunk
import java.util.*

class Location(val world: World, val position: Vector3f = Vector3f.ZERO) {

	val x: Float
		get() = this.position.x
	val y: Float
		get() = this.position.y
	val z: Float
		get() = this.position.z

	val floorX: Int
		get() = this.position.floorX
	val floorY: Int
		get() = this.position.floorY
	val floorZ: Int
		get() = this.position.floorZ

	val chunkX: Int
		get() = this.floorX shr 4
	val chunkZ: Int
		get() = this.floorZ shr 4

	val chunk: Chunk
		get() = this.world.getChunk(this.chunkX, this.chunkZ)

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is Location) return false

		return this.world == other.world && this.position == other.position
	}

	override fun hashCode() = Objects.hash(this.world, this.position)
}
